datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -2843,13 +2845,21 @@ impl Union {
                     )?
                 };
                 let nullable = fields.iter().any(|field| field.is_nullable());
-                let mut field = Field::new(name, data_type.clone(), nullable);
+
+                // Generate unique field name
+                let name = if let Some(count) = name_counts.get_mut(&base_name) {
+                    *count += 1;
```

Comment:
Make sense to me👍. But it doesn't work for this case:
```sh
> create table t1(a int);
0 row(s) fetched.
Elapsed 0.005 seconds.

> create table t2(a int);
0 row(s) fetched.
Elapsed 0.007 seconds.

> select * from t1,t2 union all select * from t1,t2;
type_coercion
caused by
Schema error: Schema contains duplicate unqualified field name
```
The above query works on the main branch.

Reply:
It seems like `TypeCoercion` undoes the renaming efforts here.
Maybe `TypeCoercion` should only modify the data types of union inputs, rather than field names 🤔

Reply:
oh right, because the union schema is discarded during coercion. My initial attempt to fix this issue propagated the union schema through, but I removed it when I decided not to maintain qualifiers, but I think adding that back would resolve this.

Reply:
Addressed in https://github.com/apache/datafusion/pull/15135/commits/612c217d13b86647b7438b6fd1054c0d8c27f766